In [3]:
pip install pandas mysql-connector-python


Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [4]:
import mysql.connector
import pandas as pd

In [5]:
connection = mysql.connector.connect(
    host="localhost",        # Your MySQL host, typically "localhost"
    user="root",             # Replace with your MySQL username (default is "root")
    password="@Nikhil85700", # Your MySQL password
    database="user_journey_data"  # Replace with the name of your database
)


In [6]:
if connection.is_connected():
    print("Successfully connected to the database!")

Successfully connected to the database!


In [7]:
query = "SELECT * FROM front_visitors;"

# Execute the query and load data into a DataFrame
df = pd.read_sql(query, connection)

# Display the first few rows of the DataFrame
print(df.head())

C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\1353736344.py:4: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


   visitor_id   user_id
0         193       NaN
1         203  274127.0
2         245  168785.0
3         277  242454.0
4         310  253829.0


In [8]:
# Query to explore the 'front_interactions' table
query = "SELECT * FROM front_interactions LIMIT 10;"
df = pd.read_sql(query, connection)
print(df.head())


   visitor_id  session_id                                   event_source_url  \
0     1076404     2801930  https://365datascience.com/login/?redirectUrl=...   
1     1076404     2801930  https://365datascience.com/login/?redirectUrl=...   
2      494069     2801929                  https://365datascience.com/login/   
3     1076404     2801930  https://365datascience.com/login/?redirectUrl=...   
4      494069     2801929                  https://365datascience.com/login/   

                               event_destination_url          event_date  \
0  https://365datascience.com/login/?redirectUrl=... 2022-12-01 00:00:07   
1  https://365datascience.com/login/?redirectUrl=... 2022-12-01 00:00:09   
2                  https://365datascience.com/login/ 2022-12-01 00:00:14   
3  https://365datascience.com/login/?redirectUrl=... 2022-12-01 00:00:14   
4                  https://365datascience.com/login/ 2022-12-01 00:00:16   

                  event_name  
0          login_email.click  


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\4269377679.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


In [9]:

query = "SELECT COUNT(*) AS total_interactions FROM front_interactions;"
df = pd.read_sql(query, connection)
print(df)


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\1318497712.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


   total_interactions
0             1625367


### Purpose
This query counts the total number of interactions recorded in the front_interactions table. Each record in this table corresponds to a unique event or interaction on the website, such as clicking a button or scrolling a page.

### Output
The query returned 1,625,367, which means there are 1,625,367 interactions recorded in the database.

### Key Insights
This gives us a baseline understanding of the scale of user activity logged on the platform.
We can use this total number to calculate percentages and analyze interaction patterns in later steps.

In [10]:
query = "SELECT COUNT(DISTINCT visitor_id) AS unique_visitors FROM front_interactions;"
df = pd.read_sql(query, connection)
print(df)


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\3312380275.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


   unique_visitors
0           245562


### Purpose
This query calculates the total number of unique visitors who interacted with the website. The DISTINCT keyword ensures that each visitor_id is only counted once, even if they have multiple interactions recorded in the front_interactions table.

### Key Insights
The count of unique visitors provides insight into the total reach of the website within the dataset's timeframe.
Comparing this number with the total number of interactions can help analyze visitor engagement. For example:
A high ratio of interactions to visitors indicates high engagement.
A low ratio suggests limited activity per visitor.


In [11]:
query = """
SELECT event_name, COUNT(*) AS event_count
FROM front_interactions
GROUP BY event_name
ORDER BY event_count DESC
LIMIT 10;
"""
df = pd.read_sql(query, connection)
print(df)


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\3248658482.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


                   event_name  event_count
0           login_email.click       145626
1   login_login.click.success       133318
2        login_password.click       108249
3           home_log_in.click        56994
4  login_google.click.success        49271
5         signup_policy.click        48343
6           signup_name.click        41694
7      login_rememberme.click        35613
8      login_login.click.fail        33956
9       signup_password.click        31985


### Purpose
This query identifies the most common event types logged in the front_interactions table. By grouping the data by event_name, we count how frequently each event type occurs. The result is ordered in descending order of event frequency to highlight the top 10 most frequent events.

### Key Insights
Most Frequent Events: The login_email.click and login_login.click.success events are the most frequent, suggesting a high level of user interaction with the login form.
Successful Login Events: The count of login_login.click.success (133,318) indicates the number of users who successfully logged in.
#### Login Issues: 
There are also 33,956 occurrences of login_login.click.fail, which may indicate login issues that warrant further investigation.
Sign-Up Interactions: Events like signup_policy.click and signup_name.click show significant activity in the user registration process.

In [12]:
query = """
SELECT event_source_url, event_destination_url, COUNT(*) AS event_count
FROM front_interactions
GROUP BY event_source_url, event_destination_url
ORDER BY event_count DESC
LIMIT 10;
"""
df = pd.read_sql(query, connection)
print(df)


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\4104561452.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


                                    event_source_url  \
0                  https://365datascience.com/login/   
1                 https://365datascience.com/signup/   
2                https://365datascience.com/courses/   
3  https://365datascience.com/login/?redirectUrl=...   
4             https://365financialanalyst.com/login/   
5                        https://365datascience.com/   
6                        https://365datascience.com/   
7  https://365datascience.com/career-track-certif...   
8                        https://365datascience.com/   
9     https://365datascience.com/course-certificate/   

                               event_destination_url  event_count  
0                  https://365datascience.com/login/       222497  
1                 https://365datascience.com/signup/       168891  
2                https://365datascience.com/courses/        78091  
3  https://365datascience.com/login/?redirectUrl=...        72272  
4             https://365financialanalyst.c

### Purpose
This query examines the relationship between source URLs (where an event originated) and destination URLs (where the event led). It provides the top 10 most frequent transitions between pages/events by counting how many interactions occurred for each source-destination pair.


### Key Insights


High Traffic Between Login and Signup Pages:

The transition from https://365datascience.com/login/ to https://365datascience.com/signup/ is the most frequent (222,497 interactions).
Similarly, many users transition back from the signup page to the login page (168,991 interactions). This suggests significant traffic between account creation and login functionalities.
Courses Page Engagement:

The https://365datascience.com/courses/ page frequently transitions to the main page (https://365datascience.com/) with 72,227 interactions. This could indicate users exploring course offerings and returning to the homepage.
Redirect Patterns:

Redirects, such as https://365datascience.com/login/?redirectUrl=, show a notable amount of traffic (59,574 interactions). This highlights the importance of analyzing redirection flows in the user journey.
Cross-Domain Traffic:

There is notable interaction between https://365financialanalyst.com/login/ and https://365datascience.com/login/ (49,234 events), indicating integration between platforms.

In [13]:
query = """
SELECT 
    COUNT(*) AS total_visitors,
    COUNT(user_id) AS total_users
FROM front_visitors;
"""
df = pd.read_sql(query, connection)
print(df)


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\2337883420.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


   total_visitors  total_users
0          245562       151759


### Purpose

This query calculates the total number of visitors and the total number of users in the front_visitors table:

Total Visitors: Represents all records in the table (COUNT(*)).
Total Users: Represents the count of non-NULL user_id values, as only visitors who created accounts are assigned a user_id.


### Key Insights
Visitor to User Conversion:

Out of 245,562 visitors, 151,759 created user accounts.
Conversion Rate: 
151
,
759
245
,
562
×
100
≈
61.8
%
245,562
151,759
​
 ×100≈61.8%
This suggests that approximately 61.8% of visitors convert into registered users.
Non-User Visitors:

### The remaining 
245, 562 −151, 759 = 93, 803
245, 562−151, 759=93, 803 visitors did not create user accounts, representing about 38.2% of the total visitors.
Potential Analysis:

Investigate the behavior of visitors who didn’t convert to users to identify potential barriers in the user journey.
Focus on improving pages or interactions where drop-offs occur.

In [14]:
query = """
SELECT COUNT(*) AS visitors_without_users
FROM front_visitors
WHERE user_id IS NULL;
"""
df = pd.read_sql(query, connection)
print(df)


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\2760915703.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


   visitors_without_users
0                   93803


### Purpose
This query calculates the number of visitors who did not create a user account. In the front_visitors table, visitors who have not registered are indicated by a NULL value in the user_id column. Counting these NULL values provides insights into the portion of visitors who visited the website but didn’t convert into registered users.

### Key Insights
Non-Converting Visitors:

Out of the total 245,562 visitors, 93,803 did not create user accounts.
This represents approximately 38.2% of the visitors, meaning a significant portion of users are not converting to registered users.
Potential Focus Area:

Investigate the behavior of these non-converting visitors by analyzing their interactions in the front_interactions table.
Identify possible barriers in the user journey, such as confusing UI, lengthy signup forms, or lack of incentive to register.

## student purchases

In [15]:
query = "SELECT COUNT(*) AS total_purchases FROM student_purchases;"
df = pd.read_sql(query, connection)
print(df)


   total_purchases
0             3503


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\2565240504.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


### Purpose
This query calculates the total number of purchases recorded in the student_purchases table. Each record in this table represents a purchase made by a user, including initial purchases and recurring payments for subscriptions.

### Key Insights
Total Purchases:

The dataset contains 3,503 total purchases made by users.
This includes all purchase types (e.g., monthly, quarterly, and annual subscriptions).

### Potential for Analysis:

Understanding the breakdown of purchase types (purchase_type) and trends over time can provide deeper insights into user purchasing behavior.
Investigating the purchase_price can help analyze the revenue generated and identify test users who made purchases at $0.

In [16]:
query = """
SELECT purchase_type, COUNT(*) AS count
FROM student_purchases
GROUP BY purchase_type;
"""
df = pd.read_sql(query, connection)
print(df)


   purchase_type  count
0              2   1909
1              1    106
2              0   1488


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\1557914434.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


### Purpose
This query provides a breakdown of the number of purchases by purchase_type in the student_purchases table. The purchase_type column categorizes purchases into different subscription types:

0: Monthly
1: Quarterly
2: Annual
##### By grouping the data by purchase_type, the query counts the number of purchases for each category.



### Key Insights
Annual Subscriptions (Type 2):

Annual subscriptions are the most popular, with 1,909 purchases, accounting for approximately 54.5% of all purchases.
Monthly Subscriptions (Type 0):

Monthly subscriptions are the second most popular, with 1,488 purchases (~42.5%).
Quarterly Subscriptions (Type 1):

Quarterly subscriptions are the least popular, with only 106 purchases, representing a small portion (~3%).
Potential User Preferences:

Users seem to prefer longer-term plans (annual) over shorter-term plans (monthly or quarterly), which could indicate a willingness to commit to the service for extended periods.

In [17]:
query = "SELECT COUNT(*) AS test_users FROM student_purchases WHERE purchase_price = 0;"
df = pd.read_sql(query, connection)
print(df)


   test_users
0           5


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\2104281557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


### Purpose
This query identifies the number of test users in the student_purchases table. Test users are defined as those who have made purchases at a price of $0, which is a common practice for testing purposes in systems.


### Key Insights
Number of Test Users:

The dataset includes 5 test users who have made purchases with a price of $0.
These records should be excluded from further revenue or user behavior analysis to ensure data accuracy.
Impact on Analysis:

Test users may skew revenue metrics and purchasing behavior analysis if not removed.
Identifying and excluding these users ensures cleaner data and more accurate insights.

In [19]:
query = """
SELECT 
    MAX(purchase_price) AS max_price,
    MIN(purchase_price) AS min_price,
    AVG(purchase_price) AS avg_price
FROM student_purchases;
"""
df = pd.read_sql(query, connection)
print(df)


   max_price  min_price  avg_price
0      348.0        0.0  84.851116


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\3789358770.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


### Purpose
This query calculates key statistics about purchase prices in the student_purchases table:

Maximum Purchase Price (MAX): The highest price paid for a purchase.
Minimum Purchase Price (MIN): The lowest price paid for a purchase.
Average Purchase Price (AVG): The average price paid across all purchases.



### Key Insights
Maximum Purchase Price:

The highest price paid for a subscription is **$348.00.** This likely represents an annual subscription or a premium plan.
Minimum Purchase Price:

The lowest price is $0.00, which corresponds to the test users identified in a previous query.
Average Purchase Price:

The average purchase price across all transactions is approximately $84.85. This value reflects the mix of subscription types (monthly, quarterly, annual) and the test user transactions.

In [21]:
query = """SELECT 
    MAX(purchase_price) AS max_price,
    MIN(purchase_price) AS min_price,
    AVG(purchase_price) AS avg_price
FROM student_purchases
WHERE purchase_price > 0;
"""

df = pd.read_sql(query,connection)
print(df)

   max_price  min_price  avg_price
0      348.0       1.16  84.972401


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\827945178.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query,connection)


### Purpose
This query recalculates the purchase price statistics, excluding test users (where purchase_price = 0). The focus is on meaningful data from actual paying users:

##### Maximum Purchase Price (MAX): The highest price paid.
##### Minimum Purchase Price (MIN): The lowest price paid (excluding $0).
##### Average Purchase Price (AVG): The average price paid by real users.

In [22]:
query = """SELECT 
    purchase_type,
    MAX(purchase_price) AS max_price,
    MIN(purchase_price) AS min_price,
    AVG(purchase_price) AS avg_price
FROM student_purchases
GROUP BY purchase_type;

"""

df = pd.read_sql(query,connection)
print(df)

   purchase_type  max_price  min_price   avg_price
0              2     348.00       0.00  129.791624
1              1     118.51       0.00   74.133774
2              0      45.72       6.08   27.959052


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\1408460327.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query,connection)


In [23]:
query = """SELECT SUM(purchase_price) AS total_revenue
FROM student_purchases
WHERE purchase_price > 0;


"""

df = pd.read_sql(query,connection)
print(df)

   total_revenue
0      297233.46


C:\Users\nurec\AppData\Local\Temp\ipykernel_6616\87190217.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query,connection)


### Purpose
This query calculates the total revenue generated from all purchases made by real users. By excluding purchases with a price of $0 (test users), the calculation reflects only meaningful revenue.

### Key Insights
#### total Revenue:

##### The total revenue generated by paying users is approximately **$297,233.46.**
This value reflects the cumulative amount paid across all subscription types (monthly, quarterly, annual).
Implications for Business:

This figure provides a snapshot of the platform's financial performance.
It can serve as a baseline for analyzing revenue trends over time or comparing the revenue contribution of different subscription types.
